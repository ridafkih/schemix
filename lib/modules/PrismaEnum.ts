import { PrismaEnumOptions, Enum } from "typings/prisma-enum";

export class PrismaEnum {
  private enumMap: Map<string, string | undefined> = new Map();

  constructor(public readonly name: string) {}

  public addValue(value: string, options?: PrismaEnumOptions) {
    this.enumMap.set(value, options?.map);
    return this;
  }

  public fromJSEnum<T extends Enum<T>>(enumLike: T) {
    const values = Object.entries(enumLike);

    values.forEach(([key, value]) => {
      if (!isNaN(parseInt(key))) return;
      const mapValue = typeof value === "string" ? value : undefined;
      this.enumMap.set(key, mapValue);
    });

    return this;
  }

  public toString() {
    const { enumMap } = this;
    const entries = [...enumMap.entries()];

    const padding = entries.reduce((accumulator: number, [{ length }]) => {
      return length > accumulator ? length : accumulator;
    }, 0);

    const lines = entries
      .map(([key, value]) => {
        if (!value) return `  ${key}`;
        return "  " + [key.padEnd(padding), ` @map("${value}")`].join(" ");
      })
      .join("\n");

    return [`enum ${this.name} {`, lines, "}"].join("\n");
  }
}
