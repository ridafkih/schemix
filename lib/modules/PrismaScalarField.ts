import {
  PrismaFieldAttribute,
  PrismaFieldModifier,
  PrismaFieldTypeName,
} from "typings/prisma-field";
import { Comment } from "typings/prisma-type-options";

export class PrismaScalarField {
  private attributes: Map<string, PrismaFieldAttribute> = new Map();
  private modifier: PrismaFieldModifier = "";
  private rawAttributeString = "";
  private comments: Comment[] = [];

  constructor(
    private readonly name: string,
    private type: PrismaFieldTypeName
  ) {}

  public setOptional() {
    this.modifier = "?";
    return this;
  }

  public setList() {
    this.modifier = "[]";
    return this;
  }

  public setAsId() {
    this.attributes.set("id", "@id");
    return this;
  }

  public setUnique() {
    this.attributes.set("unique", "@unique");
    return this;
  }

  public setToUpdatedAt() {
    this.attributes.set("updatedAt", "@updatedAt");
    return this;
  }

  public setDefault(
    defaultValue:
      | string
      | number
      | { cuid: boolean; uuid: boolean; now: boolean }
  ) {
    const setDefaultValue = (value: string) =>
      this.attributes.set("default", `@default(${value})`);
    switch (typeof defaultValue) {
      case "object":
        const [prismaFunc] =
          Object.entries(defaultValue).find(([, value]) => value) || [];
        setDefaultValue(`${prismaFunc}()`);
        break;
      case "string":
        setDefaultValue(`"${defaultValue}"`);
        break;
      default:
        setDefaultValue(defaultValue.toString());
    }

    return this;
  }

  public setIgnore() {
    this.attributes.set("ignore", "@ignore");
    return this;
  }

  public mapTo(fieldName: string) {
    this.attributes.set("map", `@map("${fieldName}")`);
    return this;
  }

  public setRawAttributes(rawAttributeString: string) {
    this.rawAttributeString = rawAttributeString;
    return this;
  }

  public setComments(comments: Comment[]) {
    this.comments = comments;
    return this;
  }

  public toFieldData() {
    return {
      comments: this.comments,
      tokens: this.toTokenArray(),
    };
  }

  public toTokenArray() {
    const { name, type, modifier, attributes, rawAttributeString } = this;
    return [
      name,
      type + modifier,
      [...attributes.values(), rawAttributeString].join(" "),
    ] as string[];
  }
}
