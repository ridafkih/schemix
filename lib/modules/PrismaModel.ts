import { PrismaEnum } from "modules/PrismaEnum";
import { PrismaSchema } from "modules/PrismaSchema";
import { PrismaEnumField } from "modules/PrismaEnumField";
import { PrismaScalarField } from "modules/PrismaScalarField";
import { PrismaRelationalField } from "modules/PrismaRelationalField";

import {
  buildCompositeId,
  buildCompositeUnique,
  buildModelMap,
  handleEnumOptions,
  handleRelationalOptions,
  handleScalarOptions,
} from "util/options";
import { parseStringOrObject } from "util/parse";

import { PrismaFieldTypeName } from "typings/prisma-field";
import {
  BooleanFieldOptions,
  CompositeIDFieldOptions,
  CompositeUniqueFieldOptions,
  DateTimeFieldOptions,
  DecimalFieldOptions,
  EnumFieldOptions,
  FieldOptions,
  FloatFieldOptions,
  IntFieldOptions,
  JsonFieldOptions,
  ModelMapOptions,
  RelationalFieldOptions,
  StringFieldOptions,
  Comment,
} from "typings/prisma-type-options";
import { PrismaEnumOptions } from "typings/prisma-enum";

export class PrismaModel {
  private fields: Map<
    string,
    PrismaScalarField | PrismaRelationalField | PrismaEnumField
  > = new Map();
  private blockAttributes: string[] = [];
  private rawFields: string[] = [];
  private comments: Comment[] = [];
  private schemaType: "model" | "view" = "model";

  private readonly schema?: PrismaSchema;
  public readonly name?: string | null;

  constructor(
    {
      name,
      schema,
      schemaType = "model",
    }: {
      name?: string | null;
      schema?: PrismaSchema;
      schemaType?: "model" | "view";
    } = {
      schemaType: "model",
    }
  ) {
    if (name !== undefined) this.name = name;
    if (schema !== undefined) this.schema = schema;
    this.schemaType = schemaType;
  }

  public comment(...comments: Comment[]) {
    this.comments.push(...comments);
    return this;
  }

  public string(fieldName: string, options?: StringFieldOptions) {
    return this.createField(fieldName, "String", options);
  }

  public int(fieldName: string, options?: IntFieldOptions) {
    return this.createField(fieldName, "Int", options);
  }

  public bigInt(fieldName: string, options?: IntFieldOptions) {
    return this.createField(fieldName, "BigInt", options);
  }

  public decimal(fieldName: string, options?: DecimalFieldOptions) {
    if (options?.precision) {
      const rawAttributes = options.raw?.split(" ") ?? [];
      rawAttributes?.push(`@database.Decimal(${options.precision.join(", ")})`);
      options.raw = rawAttributes.join(" ");
    }

    return this.createField(fieldName, "Decimal", options);
  }

  public float(fieldName: string, options?: FloatFieldOptions) {
    return this.createField(fieldName, "Float", options);
  }

  public boolean(fieldName: string, options?: BooleanFieldOptions) {
    return this.createField(fieldName, "Boolean", options);
  }

  public dateTime(fieldName: string, options?: DateTimeFieldOptions) {
    return this.createField(fieldName, "DateTime", options);
  }

  public json(fieldName: string, options?: JsonFieldOptions) {
    const parsedDefault = options?.default
      ? parseStringOrObject(options?.default)
      : undefined;

    const newOptions = parsedDefault
      ? ({ ...options, default: parsedDefault } as FieldOptions)
      : options;

    return this.createField(fieldName, "Json", newOptions);
  }

  public enum(
    fieldName: string,
    prismaEnum: PrismaEnum,
    options?: EnumFieldOptions
  ) {
    return this.createEnumField(fieldName, prismaEnum.name, options);
  }

  public relation(
    fieldName: string,
    model: PrismaModel,
    options?: RelationalFieldOptions
  ) {
    return this.createRelation(fieldName, model, options);
  }

  public id(options: CompositeIDFieldOptions) {
    this.raw(buildCompositeId(options));
    return this;
  }

  public unique(options: CompositeUniqueFieldOptions) {
    this.raw(buildCompositeUnique(options));
    return this;
  }

  public map(options: ModelMapOptions) {
    this.raw(buildModelMap(options));
    return this;
  }

  public mixin(model: PrismaModel) {
    setImmediate(() => {
      [...model.fields.entries()].map(([key, value]) =>
        this.fields.set(key, value)
      );

      this.comments.push(...model.comments);
      this.rawFields.push(...model.rawFields);
      this.blockAttributes.push(...model.blockAttributes);
    });

    return this;
  }

  public raw(fieldText: string) {
    if (fieldText.startsWith("@@")) this.blockAttributes.push(fieldText);
    else this.rawFields.push(fieldText);
    return this;
  }

  public extend(name: string) {
    const clone = this.schema
      ? this.schema.createModel(name)
      : new PrismaModel({ name });
    clone.fields = this.fields;
    clone.rawFields = this.rawFields;
    return clone;
  }

  public toString(): Promise<string> {
    return new Promise((resolve) => {
      setImmediate(() => {
        resolve(
          [
            ...this.comments,
            `${this.schemaType} ${this.name} {`,
            this.parseFields(),
            "}",
          ].join("\n")
        );
      });
    });
  }

  private createRelation(
    fieldName: string,
    model: PrismaModel,
    options: RelationalFieldOptions = {}
  ) {
    if (!model.name) return this;
    const field = new PrismaRelationalField(fieldName, model.name);
    handleRelationalOptions(field, options);
    setImmediate(() => this.fields.set(fieldName, field));
    return this;
  }

  private createEnumField(
    fieldName: string,
    type: string,
    options: PrismaEnumOptions = {}
  ) {
    const field = new PrismaEnumField(fieldName, type);
    handleEnumOptions(field, options);
    setImmediate(() => this.fields.set(fieldName, field));
    return this;
  }

  private createField(
    fieldName: string,
    type: PrismaFieldTypeName,
    options: FieldOptions = {}
  ) {
    const field = new PrismaScalarField(fieldName, type);
    handleScalarOptions(field, options);
    setImmediate(() => this.fields.set(fieldName, field));
    return this;
  }

  private parseFields() {
    const fields = [...this.fields.values()].map((field) =>
      field.toFieldData()
    );

    const mostTokens = Math.max(...fields.map(({ tokens }) => tokens.length));
    const paddings = Array(mostTokens).fill(0);

    for (let i = 0; i < mostTokens; i++)
      for (const { tokens } of fields)
        if (!tokens[i]) continue;
        else
          paddings[i] =
            tokens[i].length > paddings[i] ? tokens[i].length : paddings[i];

    return [
      ...fields.map(({ tokens, comments }) => {
        return [
          ...comments.map((comment) => "  " + comment),
          "  " +
            tokens
              .map((token, index) => token.padEnd(paddings[index]))
              .join(" ")
              .trim(),
        ].join("\n");
      }),
      ...this.rawFields.map((rawField) => "  " + rawField),
      ...(this.blockAttributes.length
        ? [
            "",
            ...this.blockAttributes.map(
              (blockAttribute) => "  " + blockAttribute
            ),
          ]
        : []),
    ].join("\n");
  }
}
