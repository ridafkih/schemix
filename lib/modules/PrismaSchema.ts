import { PrismaModel } from "modules/PrismaModel";
import { PrismaEnum } from "modules/PrismaEnum";

import { exportSchema } from "util/export";
import { parseKeyValueBlock } from "util/blocks";

import { PrismaDataSourceOptions } from "typings/prisma-datasource";
import {
  PrismaGeneratorOptions,
  PrismaMultiGeneratorOptions,
} from "typings/prisma-generator";
import { importAllFiles } from "util/import";

export class PrismaSchema {
  private enums: Map<string, PrismaEnum> = new Map();
  private models: Map<string, PrismaModel> = new Map();

  constructor(
    datasource: PrismaDataSourceOptions,
    generator: PrismaGeneratorOptions | PrismaMultiGeneratorOptions
  );
  constructor(
    datasource: PrismaDataSourceOptions,
    generator: PrismaGeneratorOptions | PrismaMultiGeneratorOptions,
    basePath: string
  );
  constructor(
    datasource: PrismaDataSourceOptions,
    generator: PrismaGeneratorOptions | PrismaMultiGeneratorOptions,
    basePath: string,
    additionalPaths: string[]
  );
  constructor(
    private readonly datasource: PrismaDataSourceOptions,
    private readonly generator:
      | PrismaGeneratorOptions
      | PrismaMultiGeneratorOptions,
    private readonly basePath?: string,
    private readonly additionalPaths?: string[]
  ) {}

  /**
   * Parses a datasource block for the schema using the parameters
   * provided in the constructor.
   * @returns A string representing the datasource block.
   */
  private parseDataSource() {
    return parseKeyValueBlock(
      "datasource",
      "database",
      Object.entries(this.datasource)
    );
  }

  /**
   * Parses a generator block for the schema using the parameters
   * provided in the constructor.
   * @returns A string representing the generator block.
   */
  private parseGenerator() {
    const generators = Array.isArray(this.generator)
      ? this.generator
      : [this.generator];

    return generators
      .map(({ name = "client", ...generator }) =>
        parseKeyValueBlock(
          "generator",
          name,
          Object.entries(generator) as [string, string | string[]][]
        )
      )
      .join("\n\n");
  }

  /**
   * Creates a mixin and automatically attaches it to the schema.
   * @returns The `PrismaModel` object.
   */
  public createMixin() {
    const model = new PrismaModel(null, this);
    return model;
  }

  /**
   * Creates a `PrismaModel` object and automatically attaches it to the schema.
   * @param modelName The name of the model.
   * @returns The `PrismaModel` object.
   */
  public createModel(modelName: string) {
    const model = new PrismaModel(modelName, this);
    this.models.set(modelName, model);
    return model;
  }

  /**
   * Creates a `PrismaEnum` object and automatically attaches it to the schema.
   * @param enumName The name of the enum.
   * @returns The `PrismaEnum` object.
   */
  public createEnum(enumName: string) {
    const prismaEnum = new PrismaEnum(enumName);
    this.enums.set(enumName, prismaEnum);
    return prismaEnum;
  }

  /**
   * Parses the schema into a singular schema string.
   * @returns Returns a singular schema string.
   */
  public toString(): Promise<string> {
    return new Promise(async (resolve) => {
      if (this.basePath) {
        await importAllFiles(this.basePath, "enums");
        await importAllFiles(this.basePath, "models");
        await importAllFiles(this.basePath, "mixins");
      }

      if (this.additionalPaths) {
        await Promise.all(
          this.additionalPaths.map(async (path) => {
            await importAllFiles(path, "enums");
            await importAllFiles(path, "models");
            await importAllFiles(path, "mixins");
          })
        );
      }

      process.nextTick(() => {
        const models = [
          this.parseDataSource(),
          this.parseGenerator(),
          ...this.enums.values(),
          ...this.models.values(),
        ];

        Promise.all(models.map((model) => model.toString())).then(
          (stringModels) => {
            const schemaString = stringModels.join("\n\n") + "\n";
            resolve(schemaString);
          }
        );
      });
    });
  }

  /**
   * Exports the schema to the provided filepath/filename.
   * @param filepath The target filepath.
   * @param filename The target filename.
   */
  public export(filepath: string, filename: string) {
    exportSchema(filepath, filename, this);
  }
}
