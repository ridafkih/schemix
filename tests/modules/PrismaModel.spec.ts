import { describe, expect, it } from "vitest";
import { PrismaModel } from "schemix/lib/modules/PrismaModel";
import { PrismaEnum } from "schemix/lib/modules/PrismaEnum";

const scalarFieldTypes = [
  "string",
  "int",
  "bigInt",
  "decimal",
  "float",
  "boolean",
  "dateTime",
  "json",
];

describe("PrismaModel", () => {
  it("Should support setting the model name", () => {
    const model = new PrismaModel("Model");
    expect(model.name).toBe("Model");
  });
  it("Should support every scalar field type", async () => {
    const model = new PrismaModel("Model").string("id", {
      default: { uuid: true },
      id: true,
    });

    for (const scalarFieldType of scalarFieldTypes)
      model[scalarFieldType](scalarFieldType);

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should support adding fields", async () => {
    const model = new PrismaModel("User")
      .string("id", {
        default: { uuid: true },
        id: true,
      })
      .string("email", { optional: true });

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should support adding unique constraints", async () => {
    const model = new PrismaModel("User")
      .string("id", { default: { uuid: true }, id: true })
      .string("email")
      .string("phoneNumber")
      .unique(["email", "phoneNumber"]);

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should support adding a map", async () => {
    const model = new PrismaModel("User")
      .string("id", { default: { uuid: true }, id: true })
      .map("normal_user");

    const asString = await model.toString();
    expect(asString).toMatchSnapshot(asString);
  });
  it("Should allow adding a composite id attribute", async () => {
    const model = new PrismaModel("User")
      .string("email")
      .id({ fields: ["email"] });

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should allow setting @updatedAt and @default(now()) on DateTime fields", async () => {
    const model = new PrismaModel("User")
      .string("id", { default: { uuid: true }, id: true })
      .dateTime("updatedAt", { updatedAt: true })
      .dateTime("createdAt", { default: { now: true } });

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should allow setting defaults for Json fields", async () => {
    const model = new PrismaModel("User")
      .string("id", { default: { uuid: true }, id: true })
      .json("json", { default: { isJson: true } });

    const asString = await model.toString();
    expect(asString).toMatchSnapshot(asString);
  });
  it("Should allow appending raw attributes", async () => {
    const model = new PrismaModel("Ambiguous")
      .string("id", { default: { uuid: true }, id: true })
      .raw('@@map("normal_user")');

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should allow appending raw fields", async () => {
    const model = new PrismaModel("Ambiguous")
      .string("id", { default: { uuid: true }, id: true })
      .raw("email String?");

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should support adding enum values", async () => {
    const enumValues = new PrismaEnum("Beans")
      .addValue("GREEN")
      .addValue("NOT_GREEN");

    const model = new PrismaModel("Farm")
      .string("id", { default: { uuid: true }, id: true })
      .enum("beans", enumValues);

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should support adding mixins", async () => {
    const uuidMixin = new PrismaModel().string("id", {
      default: { uuid: true },
      id: true,
    });

    const model = new PrismaModel("User")
      .mixin(uuidMixin)
      .string("email")
      .string("phoneNumber");

    const asString = await model.toString();
    expect(asString).toMatchSnapshot(asString);
  });
  it("Should support using raw fields with mixins", async () => {
    const emailMixin = new PrismaModel()
      .string("email")
      .raw("@@index([email])");

    const model = new PrismaModel("User")
      .string("id", { default: { uuid: true }, id: true })
      .mixin(emailMixin)
      .string("phoneNumber");

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should allow extending models", async () => {
    const base = new PrismaModel().string("id", {
      default: { uuid: true },
      id: true,
    });
    const model = base.extend("User").string("email").string("phoneNumber");

    const asString = await model.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should allow relations", async () => {
    const postModel = new PrismaModel("Post");
    const userModel = new PrismaModel("User");

    postModel
      .string("id", { default: { uuid: true }, id: true })
      .string("content")
      .relation("author", userModel, {
        fields: ["authorId"],
        references: ["id"],
        optional: true,
      })
      .string("authorId", { optional: true });

    userModel
      .string("id", { default: { uuid: true }, id: true })
      .relation("posts", postModel, { list: true });

    const postModelAsString = await postModel.toString();
    const userModelAsString = await userModel.toString();

    expect(postModelAsString).toMatchSnapshot();
    expect(userModelAsString).toMatchSnapshot();
  });
});
