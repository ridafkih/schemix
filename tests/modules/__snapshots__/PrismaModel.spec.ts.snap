// Vitest Snapshot v1

exports[`PrismaModel > Should allow adding a composite id attribute 1`] = `
"model User {
  email String

  @@id([email])
}"
`;

exports[`PrismaModel > Should allow appending raw attributes 1`] = `
"model Ambiguous {
  id String @default(uuid()) @id

  @@map(\\"normal_user\\")
}"
`;

exports[`PrismaModel > Should allow appending raw fields 1`] = `
"model Ambiguous {
  id String @default(uuid()) @id
  email String?
}"
`;

exports[`PrismaModel > Should allow extending models 1`] = `
"model User {
  id          String @default(uuid()) @id
  email       String
  phoneNumber String
}"
`;

exports[`PrismaModel > Should allow relations 1`] = `
"model Post {
  id       String  @default(uuid()) @id
  content  String
  author   User?   @relation(fields: [authorId], references: [id])
  authorId String?
}"
`;

exports[`PrismaModel > Should allow relations 2`] = `
"model User {
  id    String @default(uuid()) @id
  posts Post[]
}"
`;

exports[`PrismaModel > Should allow setting @updatedAt and @default(now()) on DateTime fields 1`] = `
"model User {
  id        String   @default(uuid()) @id
  updatedAt DateTime @updatedAt
  createdAt DateTime @default(now())
}"
`;

exports[`PrismaModel > Should allow setting defaults for Json fields > model User {
  id   String @default(uuid()) @id
  json Json   @default("{\\"isJson\\":true}")
} 1`] = `
"model User {
  id   String @default(uuid()) @id
  json Json   @default(\\"{\\\\\\"isJson\\\\\\":true}\\")
}"
`;

exports[`PrismaModel > Should support adding a map > model User {
  id String @default(uuid()) @id

  @@map("normal_user")
} 1`] = `
"model User {
  id String @default(uuid()) @id

  @@map(\\"normal_user\\")
}"
`;

exports[`PrismaModel > Should support adding enum values 1`] = `
"model Farm {
  id    String @default(uuid()) @id
  beans Beans
}"
`;

exports[`PrismaModel > Should support adding fields 1`] = `
"model User {
  id    String  @default(uuid()) @id
  email String?
}"
`;

exports[`PrismaModel > Should support adding mixins > model User {
  id          String @default(uuid()) @id
  email       String
  phoneNumber String
} 1`] = `
"model User {
  id          String @default(uuid()) @id
  email       String
  phoneNumber String
}"
`;

exports[`PrismaModel > Should support adding unique constraints 1`] = `
"model User {
  id          String @default(uuid()) @id
  email       String
  phoneNumber String

  @@unique([email, phoneNumber])
}"
`;

exports[`PrismaModel > Should support every scalar field type 1`] = `
"model Model {
  id       String   @default(uuid()) @id
  string   String
  int      Int
  bigInt   BigInt
  decimal  Decimal
  float    Float
  boolean  Boolean
  dateTime DateTime
  json     Json
}"
`;

exports[`PrismaModel > Should support using raw fields with mixins 1`] = `
"model User {
  id          String @default(uuid()) @id
  email       String
  phoneNumber String

  @@index([email])
}"
`;
